### PR TITLE
MIG-284: add phase 1 progress reporting to dvm controller

### PR DESCRIPTION
**Pull request by alaypatel07**
This adds a new CR `DirectVolumeMigrationProgress`. 

The DirectVolumeMigrationProgress polls for the rsync-client
pod on source cluster for 5 seconds. It reports the state of the
pod as running, failed, or successful. 

If it is able to find a pod in running state in that 5 seconds, 
it will report the last 5 lines of the logs. 

It will then try to parse and get progress percent and transfer 
speed.

All the data in the status is accumulated at 5-second interval.
It is likely that this controller will miss the pods in action.
Hence the DirectVolumeMigration Controller should query the pods
directly to enquire for status and _not rely_ on this controller.

The migmigration controller should rely on this controller for progress
information if the status of this controller indicated `PodRunning` to 
disperse progress information to the UI

### CODE REVIEW

**Comment by alaypatel07** [comment]
```
status:
  conditions:
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-11-03T22:00:14Z"
    message: 'The migration has failed.  See: Errors.'
    reason: WaitForRsyncClientPodsCompleted
    status: "True"
    type: Failed
  errors:
  - One or more pods are in error state
  failedPods:
  - name: directvolumemigration-rsync-transfer-pvc-0
    namespace: pvc-migrate-bmark0
  itinerary: Failed
  observedDigest: fd9870dc12dde5f480b520377a9b5f6b6984fc0d2378dcb94f1ab721557731b2
  phase: Completed
  startTimestamp: "2020-11-03T21:57:50Z"

$ oc get directpvmigrationprogress -n pvc-migrate-bmark0     directvolumemigration-rsync-transfer-pvc-0 -oyaml
..
..
..
status:
  conditions:
  - category: Required
    lastTransitionTime: "2020-11-03T21:16:00Z"
    message: The progress is available
    status: "True"
    type: Ready
  logMessage: |
    rsync: --human: unknown option
    rsync error: syntax or usage error (code 1) at main.c(1578) [client=3.1.3]
  observedDigest: 419049d2080899448c62e94767b4fa9dac4a055e134c86ea1e740ace3a489f7a
  phase: Failed
```

With the most recent change added the logic to:
- fail the migration to if one or more failed rsync client pods.
- see which pods failed on dvm CR
- get the dpmp cr to see `logMessage`on why the Pod was failing

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+// ReconcileDirectPVMigrationProgress reconciles a DirectPVMigrationProgress object
+type ReconcileDirectPVMigrationProgress struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a DirectPVMigrationProgress object and makes changes based on the state read
+// and what is in the DirectPVMigrationProgress.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  The scaffolding writes
```
Comment (dymurray) [inline pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go:109]: Remove

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+	}
+
+	if !pvProgress.Status.IsReady() {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// we will requeue this every 5 seconds
+	return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 5}, nil
+
+	// at this point assume that the container is running
```
Comment (dymurray) [inline pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go:171]: Remove?

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+	return parseLogs(readCloser)
+}
+
+func (r *ReconcileDirectPVMigrationProgress) GetProgressPercent(message string) string {
+	return GetLastMatch(`\d+\%`, message)
+}
+
+func (r *ReconcileDirectPVMigrationProgress) GetTransferRate(message string) string {
+	// Transfer Rate
+	return GetLastMatch(`\d+\.\w*\/s`, message)
```
Comment (dymurray) [inline pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go:354]: Nice :)

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_suite_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	//t := &envtest.Environment{
```
Comment (dymurray) [inline pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_suite_test.go:35]: Remove?

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	if apierrors.IsInvalid(err) {
+		t.Logf("failed to create object, got an invalid object error: %v", err)
+		return
+	}
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer c.Delete(context.TODO(), instance)
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
+}
+
+func Test_parseLogs(t *testing.T) {
```
Comment (dymurray) [inline pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_test.go:78]: +1

File: pkg/controller/directvolumemigration/rsync.go [hunk 6]
```diff
@@ -603,15 +604,25 @@ func (t *Task) createRsyncClientPods() error {
 				},
 			})
 			containers = append(containers, corev1.Container{
-				Name:  fmt.Sprintf("rsync-client-%s", vol),
+				Name:  "rsync-client",
```
Comment (dymurray) [inline pkg/controller/directvolumemigration/rsync.go:607]: Thanks this was left over from when we were doing all the clients in multiple containers one pod

File: pkg/controller/directvolumemigration/rsync.go [hunk 7]
```diff
@@ -603,15 +604,25 @@ func (t *Task) createRsyncClientPods() error {
...
 				Image: "quay.io/konveyor/rsync-transfer:latest",
 				Env: []corev1.EnvVar{
 					{
 						Name:  "RSYNC_PASSWORD",
 						Value: "changeme",
 					},
 				},
-				Command: []string{"rsync", "-aPvvHh", "--delete", "--port", "2222", "--log-file", "/dev/stdout", fmt.Sprintf("/mnt/%s/%s/", ns, vol), fmt.Sprintf("rsync://root@%s/%s", ip, vol)},
+				TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+				Command: []string{"rsync",
```
Comment (dymurray) [inline pkg/controller/directvolumemigration/rsync.go:616]: +1

File: pkg/controller/directvolumemigration/rsync.go [hunk 8]
```diff
@@ -708,6 +781,15 @@ func (t *Task) deleteRsyncResources() error {
 		return err
 	}
 
+	if !t.Owner.Spec.DeleteProgressReportingCRs {
+		return nil
+	}
+
+	err = t.deleteProgressReportingDRs(destClient)
```
Comment (dymurray) [inline pkg/controller/directvolumemigration/rsync.go:788]: Why `deleteProgressReportingDRs`? Shouldn't it be `CRs`?

File: pkg/controller/directvolumemigration/rsync.go [hunk 9]
```diff
@@ -825,3 +907,22 @@ func (t *Task) findAndDeleteResources(client compat.Client) error {
 	}
 	return nil
 }
+
+func (t *Task) deleteProgressReportingDRs(client compat.Client) error {
```
Comment (dymurray) [inline pkg/controller/directvolumemigration/rsync.go:911]: `deleteProgressReportingCRs`?

**Review by dymurray — APPROVED** [review]

File: pkg/controller/directvolumemigration/rsync.go (hunk 8 above)
Comment (alaypatel07) [reply pkg/controller/directvolumemigration/rsync.go:788]: yes, I must be in a really weird state to make that kind of a typo HAHA

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_suite_test.go (hunk 4 above)
Comment (alaypatel07) [reply pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_suite_test.go:35]: that was my local hack to make unit test work, I will uncomment them

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go (hunk 2 above)
Comment (alaypatel07) [reply pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller.go]: done

File: pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_suite_test.go (hunk 4 above)
Comment (alaypatel07) [reply pkg/controller/directpvmigrationprogress/directpvmigrationprogress_controller_suite_test.go]: done

File: pkg/controller/directvolumemigration/rsync.go (hunk 8 above)
Comment (alaypatel07) [reply pkg/controller/directvolumemigration/rsync.go]: done

File: pkg/controller/directvolumemigration/rsync.go (hunk 9 above)
Comment (alaypatel07) [reply pkg/controller/directvolumemigration/rsync.go]: done

**Comment by pranavgaikwad** [comment]
@alaypatel07 can we create DMVP object in the openshift-migration namespace?

**Comment by alaypatel07** [comment]
@pranavgaikwad sure!

> can we create DMVP CR in the openshift-migration namespace?

File: config/samples/migration_v1alpha1_directpvmigrationprogress.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+apiVersion: migration.openshift.io/v1alpha1
+kind: DirectVolumeMigrationProgress
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: directpvmigrationprogress-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (shawn-hurley) [inline config/samples/migration_v1alpha1_directpvmigrationprogress.yaml:9]: probably want to at least remove this

File: pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+			Status:   migapi.True,
+			Reason:   NotFound,
+			Category: migapi.Critical,
+			Message: fmt.Sprintf("The spec.podRef %s must reference a `Pod` with container name %s",
+				path.Join(podRef.Namespace, podRef.Name), containerName),
+		})
+		return nil
+	}
+
+	switch {
```
Comment (shawn-hurley) [inline pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go:251]: need to handle waiting status as a followup I think

File: pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+		pvProgress.Status.PodPhase = kapi.PodFailed
+		// TODO: report failure
+	case !containerStatus.Ready && containerStatus.LastTerminationState.Terminated != nil && containerStatus.LastTerminationState.Terminated.ExitCode == 0:
+		// succeeded dont ever requeue
+		pvProgress.Status.PodPhase = kapi.PodSucceeded
+	case pod.Status.Phase == kapi.PodSucceeded:
+		// Its possible for the succeeded pod to not have containerStatuses at all
+		pvProgress.Status.PodPhase = kapi.PodSucceeded
+	case pod.Status.Phase == kapi.PodFailed:
+		// Its possible for the succeeded pod to not have containerStatuses at all
```
Comment (shawn-hurley) [inline pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go:280]: nit: update comment here to discuss pod failed

File: pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+
+func parseLogs(reader io.Reader) (string, error) {
+	buf := new(strings.Builder)
+	_, err := io.Copy(buf, reader)
+	if err != nil {
+		return "", err
+	}
+	data := strings.Split(buf.String(), "\n")
+	logLines := []string{}
+	for _, line := range data {
```
Comment (shawn-hurley) [inline pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go:357]: if we transfer GB's of small files, won't this be a very large list? are we worried about exceeding 1.5 MB limit?

Comment (alaypatel07) [reply pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go:357]: I think not https://github.com/konveyor/mig-controller/pull/734/files#diff-7e9c060a79e5cc19929659bc606b8ae646f8751e65a54f2a9de0392d46bf25b7R255 Just grabbing the last 5 lines.

I initially thought I would be grabbing 10 lines, but looks kube does not like me sending 10 line of chunk to a CR so it did not respect the \n and escaped them.

Having 5 lines got me to the point that the kubectl get is like this
```
  status:
    conditions:
    - category: Critical
      lastTransitionTime: "2020-11-09T19:30:28Z"
      message: 'The spec.podRef pvc-migrate-bmark0/directvolumemigration-rsync-transfer-pvc-0
        must reference a valid `Pod` '
      reason: NotFound
      status: "True"
      type: InvalidPod
    lastObservedProgressPercent: 96%
    lastObservedTransferRate: 40.23MB/s
    logMessage: |-
      1.41G  94%   40.22MB/s    0:00:33 (xfr#82, to-chk=18/101)202
      1.41G  95%   40.22MB/s    0:00:33 (xfr#83, to-chk=17/101)202
      1.42G  95%   40.23MB/s    0:00:33 (xfr#84, to-chk=16/101)202
      1.42G  95%   40.23MB/s    0:00:33 (xfr#85, to-chk=15/101)202
      1.43G  96%   40.23MB/s    0:00:33 (xfr#86, to-chk=14/101)202
    observedDigest: 419049d2080899448c62e94767b4fa9dac4a055e134c86ea1e740ace3a489f7a
```

File: pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go (hunk 11 above)
Comment (alaypatel07) [reply pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go:251]: ack

**Pull request merged**